src/test/java/com/casadocodigo/site/cadastrocategoria/CategoriaControllerTest.java

```diff
@@ -0,0 +1,73 @@
+package com.casadocodigo.site.cadastrocategoria;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.NullSource;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.web.servlet.MockMvc;
+
+import static org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@SpringBootTest
+@AutoConfigureMockMvc
+@AutoConfigureTestDatabase
```

Comment:
Dá pra fazer esse teste apenas com o `@WebMvc(CategoriaController.class)`. Um teste desse tipo não sobe nada que não seja Web. Portanto, não tem nada de acesso a BD.

Seria como um teste da API, mas isolado dos pontos de integração com o BD. Todo acesso a dados teria que ser mockado.

```suggestion
@WebMvc(CategoriaController.class)
```

Reply:
Opa, aqui eu tinha tentando subir um teste mais integrado da aplicação e subir um DB em memória para rodar junto. Queria deixar tudo mais próximo de como a aplicação será realmente utilizada.

Reply:
É uma alternativa. Vale avaliar os prós e contras.

Uma coisa do `@WebMvcTest` com Mocks é que dá pra verificar a interação de maneira fácil quando deveria e, especialmente, quando NÃO deveria chamar.

As respostas pré-fabricadas podem simplificar a carga de dados inicial.

Se todo o comportamento menos o do repository, poderíamos testar o repository pra complementar.


Um teste mais integrado realmente exercitaria mais passos da stack. Mas deixaria mais difícil verificar essas interações que fazemos com um Mock.

Seria interessante medir o tempo de execução dos dois tipos de teste, pensando não em números absolutos, mas em porcentagens. Os números absolutos são muito pequenos, se consideramos a quantidade de testes que temos.

Reply:
Realmente, quando eu faço um teste mais integrado eu não consigo fazer asserções se o dado foi realmente salvo no banco de dados e o setup do teste em si pode ser bem mais trabalhoso de ser criado. Talvez valha a pena estudar um misto de ambos os testes e conseguir derivar um tipo de teste que seja satisfatório para o sistema.

Reply:
> Se todo o comportamento menos o do repository, poderíamos testar o repository pra complementar.

Não consegui compreender muito bem o que quis ser dito :/. Poderia discorrer mais?

Reply:
> Adicionei os testes mais integrados da classe CategoriaController. Será que rolaria em um breve futuro criar uns testes automatizados com Property Based Testing para testar o fluxo da criação da categoria?

Opa, Henrique. A gente estava brincando com Property Based Tests com o jqwik outro dia. Era num projeto de treino, mas fizemos pareando via Code Share do IntelliJ e o dev não pushou a branch.

A gente fez testes de validação, configurando `Arbitraries` para valores inválidos. Ficou legal.

O plugin do jqwik para Spring Boot não funfa com `@WebMvcTest`, só com `@SpringBootTest`.

Reply:
> > Se todo o comportamento menos o do repository, poderíamos testar o repository pra complementar.
> 
> Não consegui compreender muito bem o que quis ser dito :/. Poderia discorrer mais?

Era pra ser:

"Se **quero testar** todo o comportamento menos o do repository, poderíamos testar o repository pra complementar."

Vê se dá pra sacar essa imagem:
https://drive.google.com/file/d/1_jW2ldVV9f4EzBiIYDwt_Kx6SRTte15R/view?usp=sharing


Pra esclarecer melhor, precisamos de uma _call_. hehe